### ISSUE TITLE
Incrementally adding new languages to pre-trained models

### ISSUE THREAD

**Issue by hadyelsahar** [bigscience-workshop/multilingual-modeling#2]
# Experiments design
Follow discussion [here](https://docs.google.com/document/d/110tlidAcpiNteKnA27tR5KPS_VahNqYKqCeJlu1MWww/edit#heading=h.wmf5tyes1tfk)

### Pretrained Models 
To allow comparable preplexity values across languages. We need multilingual or monolingual models that are loaded with multilingual vocabulary.
- ... 

# Datasets
- Oscar dataset tokenized using GPT tokenizer
- ....

## Code 
- Megatron-Deepspeed finetuning [see examples here](https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/examples/pretrain_gpt_tiny.sh)
```
We've provided several scripts for pretraining both BERT and GPT in examples directory, 
as well as scripts for both zero-shot and fine-tuned downstream tasks including MNLI, RACE, WikiText103, and LAMBADA evaluation. There is also a script for GPT interactive text generation. 
```
- ... 

# Progress / Results

**Comment by yongzx** [comment]
I would like to help implement the experiments.

**Comment by vnikouliNLE** [comment]
This is quite a large task. For the moment implementation-wise I see 2 subtasks: 

- [ ] We need an implementation where instead of full finetuning we only finetune specific part of the architecture: eg. input and/or output embedding, normalization, other?
- [ ]  Implement adapter layers 

Experiment-wise

 - [ ] adapt English checkpoint to French ( we can take French part of Oscar?) by finetuning embedding layer
 - [ ] adapt English checkpoint to Korean by finetuning embedding layer
 - [ ] evaluate multilingual checkpoint to Russian/Japanese/other? 
 - [ ] adapt mutilingual checkpoint to new languages (tbd)

**Comment by yongzx** [comment]
For the task of implementing adapter layers, [Adapter Hub](https://adapterhub.ml/explore/text_lang/) has pre-trained multilingual adapters on Wikipedia (Amharic, Arabic, Armenian, Basque, Bengali, Bhojpuri, Burmese, Buryat, Cantonese, Chinese, Czech, English, Erzya, Estonian, Finnish, French, Georgian, German, Greek, Guarani, Haitian Creole, Hindi, Hungarian, Icelandic, Ilocano, Indonesian, Italian, Japanese, Javanese, Komi, Korean, Latin, Latvian, Meadow Mari, Min Dong, Mingrelian, Maori, Northern Sami, Persian, Portuguese, Quechua, Russian, Spanish, Swahili, Tamil, Thai, Turkish, Turkmen, Vietnamese, Wolof). 

Referring to [Modeling Multilinguality document](https://docs.google.com/document/d/110tlidAcpiNteKnA27tR5KPS_VahNqYKqCeJlu1MWww/edit), Catalan, Urdu, and Bangla are not included in the pretrained adapters on Adapter-Hub.

**Comment by sbmaruf** [comment]
Hi! @yongzx!
Great stuffs.

All these adapters are for `bert-base-multilingual-cased`.
Hopefully we will be working with `gpt-2` and later adapt our model following this tutorial. 

https://github.com/Adapter-Hub/adapter-transformers/blob/master/adding_adapters_to_a_model.md

I see there are already example codes to train an adapter for LM here, 

https://github.com/Adapter-Hub/adapter-transformers/blob/master/examples/language-modeling/run_clm.py

Probably we will have to change some stuffs here and there. But I think we are ready to start experiments for adapters. Let's discuss about these details in the next meeting.

@vnikouliNLE

**Comment by yongzx** [comment]
Update - Dec 15, 2021 - We have results for fine-tuning GPT2 embedding layers for German and Korean. We've evaluated it on [XNLI](https://docs.google.com/spreadsheets/d/1d59uze3QYv8KQAZtJDY73HTd7q66fr0JQqNcjBeIKNU/edit#gid=483823299), and German's model has comparable performance as mBERT. 

Next Steps:
- Supervised and 0-Shot Evaluation on XQuAD and PAWS-X
- Language Adapters + Finetuned Embedding Layers
